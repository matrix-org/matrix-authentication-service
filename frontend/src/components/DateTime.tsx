// Copyright 2022 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import {
  formatISO,
  intlFormat,
  intlFormatDistance,
  differenceInHours,
  parseISO,
} from "date-fns";

type Props = {
  className?: string;
  datetime: Date | string;
  now?: Date;
};

export const formatDate = (datetime: Date): string =>
  intlFormat(datetime, {
    year: "numeric",
    month: "short",
    day: "numeric",
    weekday: "short",
    hour: "numeric",
    minute: "numeric",
  });

/**
 * Formats a datetime
 * Uses distance when less than an hour ago
 * Else internationalised `Fri, 21 Jul 2023, 16:14`
 */
export const formatReadableDate = (datetime: Date, now: Date): string =>
  Math.abs(differenceInHours(now, datetime, { roundingMethod: "round" })) > 1
    ? formatDate(datetime)
    : intlFormatDistance(datetime, now);

const DateTime: React.FC<Props> = ({
  datetime: datetimeProps,
  now: nowProps,
  className,
}) => {
  const datetime =
    typeof datetimeProps === "string" ? parseISO(datetimeProps) : datetimeProps;
  const now = nowProps || new Date();
  const text = formatReadableDate(datetime, now);

  return (
    <time className={className} dateTime={formatISO(datetime)}>
      {text}
    </time>
  );
};

export default DateTime;
