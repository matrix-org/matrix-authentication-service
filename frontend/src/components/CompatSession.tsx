// Copyright 2022 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { parseISO } from "date-fns";
import { useTranslation } from "react-i18next";
import { useMutation } from "urql";

import { FragmentType, graphql, useFragment } from "../gql";
import { DeviceType } from "../gql/graphql";

import DateTime from "./DateTime";
import EndSessionButton from "./Session/EndSessionButton";
import LastActive from "./Session/LastActive";
import * as Card from "./SessionCard";

export const FRAGMENT = graphql(/* GraphQL */ `
  fragment CompatSession_session on CompatSession {
    id
    createdAt
    deviceId
    finishedAt
    lastActiveIp
    lastActiveAt
    userAgent {
      name
      os
      model
      deviceType
    }
    ssoLogin {
      id
      redirectUri
    }
  }
`);

export const END_SESSION_MUTATION = graphql(/* GraphQL */ `
  mutation EndCompatSession($id: ID!) {
    endCompatSession(input: { compatSessionId: $id }) {
      status
      compatSession {
        id
        finishedAt
      }
    }
  }
`);

export const simplifyUrl = (url: string): string => {
  let parsed: URL;
  try {
    parsed = new URL(url);
  } catch (e) {
    // Not a valid URL, return the original
    return url;
  }

  // Clear out the search params and hash
  parsed.search = "";
  parsed.hash = "";

  if (parsed.protocol === "https:") {
    return parsed.hostname;
  }

  // Return the simplified URL
  return parsed.toString();
};

const CompatSession: React.FC<{
  session: FragmentType<typeof FRAGMENT>;
}> = ({ session }) => {
  const { t } = useTranslation();
  const data = useFragment(FRAGMENT, session);
  const [, endCompatSession] = useMutation(END_SESSION_MUTATION);

  const onSessionEnd = async (): Promise<void> => {
    await endCompatSession({ id: data.id });
  };

  const clientName = data.ssoLogin?.redirectUri
    ? simplifyUrl(data.ssoLogin.redirectUri)
    : undefined;

  const deviceType = data.userAgent?.deviceType ?? DeviceType.Unknown;

  const deviceName =
    data.userAgent?.model ??
    (data.userAgent?.name
      ? data.userAgent?.os
        ? t("frontend.session.name_for_platform", {
            name: data.userAgent.name,
            platform: data.userAgent.os,
          })
        : data.userAgent.name
      : t("frontend.session.unknown_device"));

  const createdAt = parseISO(data.createdAt);
  const lastActiveAt = data.lastActiveAt
    ? parseISO(data.lastActiveAt)
    : undefined;

  return (
    <Card.Root>
      <Card.LinkBody
        to="/sessions/$id"
        params={{ id: data.id }}
        disabled={!!data.finishedAt}
      >
        <Card.Header type={deviceType}>
          <Card.Name name={deviceName} />
          {clientName && <Card.Client name={clientName} />}
        </Card.Header>

        <Card.Metadata>
          {lastActiveAt && (
            <Card.Info label={t("frontend.session.last_active_label")}>
              <LastActive lastActive={lastActiveAt} />
            </Card.Info>
          )}
          <Card.Info label={t("frontend.session.signed_in_label")}>
            <DateTime datetime={createdAt} />
          </Card.Info>
          <Card.Info label={t("frontend.session.device_id_label")}>
            {data.deviceId}
          </Card.Info>
        </Card.Metadata>
      </Card.LinkBody>

      {!data.finishedAt && (
        <Card.Action>
          <EndSessionButton endSession={onSessionEnd}>
            <Card.Body compact>
              <Card.Header type={deviceType}>
                <Card.Name name={deviceName} />
                {clientName && <Card.Client name={clientName} />}
              </Card.Header>
            </Card.Body>
          </EndSessionButton>
        </Card.Action>
      )}
    </Card.Root>
  );
};

export default CompatSession;
