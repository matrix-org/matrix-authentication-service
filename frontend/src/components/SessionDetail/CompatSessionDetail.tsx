// Copyright 2022 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { parseISO } from "date-fns";
import { useTranslation } from "react-i18next";
import { useMutation } from "urql";

import { FragmentType, graphql, useFragment } from "../../gql";
import BlockList from "../BlockList/BlockList";
import { END_SESSION_MUTATION, simplifyUrl } from "../CompatSession";
import DateTime from "../DateTime";
import ExternalLink from "../ExternalLink/ExternalLink";
import EndSessionButton from "../Session/EndSessionButton";
import LastActive from "../Session/LastActive";

import SessionDetails from "./SessionDetails";
import SessionHeader from "./SessionHeader";

export const FRAGMENT = graphql(/* GraphQL */ `
  fragment CompatSession_detail on CompatSession {
    id
    createdAt
    deviceId
    finishedAt
    lastActiveIp
    lastActiveAt
    ssoLogin {
      id
      redirectUri
    }
  }
`);

type Props = {
  session: FragmentType<typeof FRAGMENT>;
};

const CompatSessionDetail: React.FC<Props> = ({ session }) => {
  const data = useFragment(FRAGMENT, session);
  const [, endSession] = useMutation(END_SESSION_MUTATION);
  const { t } = useTranslation();

  const onSessionEnd = async (): Promise<void> => {
    await endSession({ id: data.id });
  };

  const finishedAt = data.finishedAt
    ? [
        {
          label: t("frontend.session.finished_label"),
          value: <DateTime datetime={parseISO(data.finishedAt)} />,
        },
      ]
    : [];

  const lastActiveIp = data.lastActiveIp
    ? [
        {
          label: t("frontend.session.ip_label"),
          value: <code>{data.lastActiveIp}</code>,
        },
      ]
    : [];

  const lastActiveAt = data.lastActiveAt
    ? [
        {
          label: t("frontend.session.last_active_label"),
          value: <LastActive lastActive={parseISO(data.lastActiveAt)} />,
        },
      ]
    : [];

  const sessionDetails = [
    { label: t("frontend.session.id_label"), value: <code>{data.id}</code> },
    {
      label: t("frontend.session.device_id_label"),
      value: <code>{data.deviceId}</code>,
    },
    {
      label: t("frontend.session.signed_in_label"),
      value: <DateTime datetime={parseISO(data.createdAt)} />,
    },
    ...finishedAt,
    ...lastActiveAt,
    ...lastActiveIp,
  ];

  const clientDetails: { label: string; value: string | JSX.Element }[] = [];

  if (data.ssoLogin?.redirectUri) {
    clientDetails.push({
      label: t("frontend.compat_session_detail.name"),
      value: simplifyUrl(data.ssoLogin.redirectUri),
    });
    clientDetails.push({
      label: t("frontend.session.uri_label"),
      value: (
        <ExternalLink target="_blank" href={data.ssoLogin?.redirectUri}>
          {data.ssoLogin?.redirectUri}
        </ExternalLink>
      ),
    });
  }

  return (
    <BlockList>
      <SessionHeader to="/sessions">{data.deviceId || data.id}</SessionHeader>
      <SessionDetails
        title={t("frontend.compat_session_detail.session_details_title")}
        details={sessionDetails}
      />
      {clientDetails.length > 0 ? (
        <SessionDetails
          title={t("frontend.compat_session_detail.client_details_title")}
          details={clientDetails}
        />
      ) : null}
      {!data.finishedAt && <EndSessionButton endSession={onSessionEnd} />}
    </BlockList>
  );
};

export default CompatSessionDetail;
