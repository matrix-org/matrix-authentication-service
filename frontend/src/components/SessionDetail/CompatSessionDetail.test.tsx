// Copyright 2023 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// @vitest-environment happy-dom

import { render, cleanup } from "@testing-library/react";
import { describe, expect, it, afterEach, vi } from "vitest";

import { makeFragmentData } from "../../gql/fragment-masking";
import { COMPAT_SESSION_FRAGMENT } from "../CompatSession";
import DateTime from "../DateTime";

import CompatSessionDetail from "./CompatSessionDetail";

// Mock out datetime to avoid timezones/date formatting
vi.mock("../DateTime", () => {
  const MockDateTime: typeof DateTime = ({ datetime }) => (
    <code>{datetime.toString()}</code>
  );
  return { default: MockDateTime };
});

describe("<CompatSessionDetail>", () => {
  const baseSession = {
    id: "session-id",
    deviceId: "abcd1234",
    createdAt: "2023-06-29T03:35:17.451292+00:00",
    ssoLogin: {
      id: "test-id",
      redirectUri: "https://element.io",
    },
  };
  afterEach(cleanup);

  it("renders a compatability session details", () => {
    const data = makeFragmentData(baseSession, COMPAT_SESSION_FRAGMENT);

    const { container } = render(<CompatSessionDetail session={data} />);

    expect(container).toMatchSnapshot();
  });

  it("renders a compatability session without an ssoLogin redirectUri", () => {
    const data = makeFragmentData(
      {
        ...baseSession,
        ssoLogin: {
          id: "dfsdjfdk",
          redirectUri: undefined,
        },
      },
      COMPAT_SESSION_FRAGMENT,
    );

    const { container } = render(<CompatSessionDetail session={data} />);

    expect(container).toMatchSnapshot();
  });

  it("renders a finished compatability session details", () => {
    const data = makeFragmentData(
      {
        ...baseSession,
        finishedAt: "2023-07-29T03:35:17.451292+00:00",
      },
      COMPAT_SESSION_FRAGMENT,
    );

    const { getByText, queryByText } = render(
      <CompatSessionDetail session={data} />,
    );

    expect(getByText("Finished")).toBeTruthy();
    // no end session button
    expect(queryByText("End session")).toBeFalsy();
  });
});
