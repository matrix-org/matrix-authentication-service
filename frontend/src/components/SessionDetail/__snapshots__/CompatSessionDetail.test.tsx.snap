// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<CompatSessionDetail> > renders a compatability session details 1`] = `
<div>
  <div
    class="_blockList_f8cc7f"
  >
    <header
      class="_header_92353c"
    >
      <button
        class="_icon-button_9uxnz_17"
        style="--cpd-icon-button-size: 32px;"
        type="button"
      >
        <svg
          class="cpd-icon"
          fill="none"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12.2071 5.29289C12.5976 5.68342 12.5976 6.31658 12.2071 6.70711L7.91421 11H18.5C19.0523 11 19.5 11.4477 19.5 12C19.5 12.5523 19.0523 13 18.5 13H7.91421L12.2071 17.2929C12.5976 17.6834 12.5976 18.3166 12.2071 18.7071C11.8166 19.0976 11.1834 19.0976 10.7929 18.7071L4.79289 12.7071C4.40237 12.3166 4.40237 11.6834 4.79289 11.2929L10.7929 5.29289C11.1834 4.90237 11.8166 4.90237 12.2071 5.29289Z"
            fill="currentColor"
          />
        </svg>
      </button>
      <h3
        class="_font-heading-md-semibold_1jx6b_121"
      >
        abcd1234
      </h3>
    </header>
    <div
      class="_block_17898c"
    >
      <h6
        class="_font-body-md-semibold_1jx6b_64"
      >
        Session
      </h6>
      <ul
        class="_list_040867"
      >
        <li
          class="_detailRow_040867"
        >
          <p
            class="_font-body-sm-semibold_1jx6b_45 _detailLabel_040867"
          >
            ID
          </p>
          <p
            class="_font-body-sm-regular_1jx6b_40 _detailValue_040867"
          >
            <code>
              session-id
            </code>
          </p>
        </li>
        <li
          class="_detailRow_040867"
        >
          <p
            class="_font-body-sm-semibold_1jx6b_45 _detailLabel_040867"
          >
            Device ID
          </p>
          <p
            class="_font-body-sm-regular_1jx6b_40 _detailValue_040867"
          >
            <code>
              abcd1234
            </code>
          </p>
        </li>
        <li
          class="_detailRow_040867"
        >
          <p
            class="_font-body-sm-semibold_1jx6b_45 _detailLabel_040867"
          >
            Signed in
          </p>
          <p
            class="_font-body-sm-regular_1jx6b_40 _detailValue_040867"
          >
            <time
              datetime="2023-06-29T03:35:17Z"
            >
              Thu, 29 Jun 2023, 03:35
            </time>
          </p>
        </li>
      </ul>
    </div>
    <div
      class="_block_17898c"
    >
      <h6
        class="_font-body-md-semibold_1jx6b_64"
      >
        Client
      </h6>
      <ul
        class="_list_040867"
      >
        <li
          class="_detailRow_040867"
        >
          <p
            class="_font-body-sm-semibold_1jx6b_45 _detailLabel_040867"
          >
            Name
          </p>
          <p
            class="_font-body-sm-regular_1jx6b_40 _detailValue_040867"
          >
            element.io
          </p>
        </li>
        <li
          class="_detailRow_040867"
        >
          <p
            class="_font-body-sm-semibold_1jx6b_45 _detailLabel_040867"
          >
            Uri
          </p>
          <p
            class="_font-body-sm-regular_1jx6b_40 _detailValue_040867"
          >
            <a
              class="_link_1mzip_17 _externalLink_a97355"
              data-kind="primary"
              href="https://element.io"
              rel="noreferrer noopener"
              target="_blank"
            >
              https://element.io
            </a>
          </p>
        </li>
      </ul>
    </div>
    <button
      aria-controls="radix-:r0:"
      aria-expanded="false"
      aria-haspopup="dialog"
      class="_button_1ukgx_17"
      data-kind="destructive"
      data-size="sm"
      data-state="closed"
      role="button"
      tabindex="0"
      type="button"
    >
      End session
    </button>
  </div>
</div>
`;

exports[`<CompatSessionDetail> > renders a compatability session without an ssoLogin redirectUri 1`] = `
<div>
  <div
    class="_blockList_f8cc7f"
  >
    <header
      class="_header_92353c"
    >
      <button
        class="_icon-button_9uxnz_17"
        style="--cpd-icon-button-size: 32px;"
        type="button"
      >
        <svg
          class="cpd-icon"
          fill="none"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12.2071 5.29289C12.5976 5.68342 12.5976 6.31658 12.2071 6.70711L7.91421 11H18.5C19.0523 11 19.5 11.4477 19.5 12C19.5 12.5523 19.0523 13 18.5 13H7.91421L12.2071 17.2929C12.5976 17.6834 12.5976 18.3166 12.2071 18.7071C11.8166 19.0976 11.1834 19.0976 10.7929 18.7071L4.79289 12.7071C4.40237 12.3166 4.40237 11.6834 4.79289 11.2929L10.7929 5.29289C11.1834 4.90237 11.8166 4.90237 12.2071 5.29289Z"
            fill="currentColor"
          />
        </svg>
      </button>
      <h3
        class="_font-heading-md-semibold_1jx6b_121"
      >
        abcd1234
      </h3>
    </header>
    <div
      class="_block_17898c"
    >
      <h6
        class="_font-body-md-semibold_1jx6b_64"
      >
        Session
      </h6>
      <ul
        class="_list_040867"
      >
        <li
          class="_detailRow_040867"
        >
          <p
            class="_font-body-sm-semibold_1jx6b_45 _detailLabel_040867"
          >
            ID
          </p>
          <p
            class="_font-body-sm-regular_1jx6b_40 _detailValue_040867"
          >
            <code>
              session-id
            </code>
          </p>
        </li>
        <li
          class="_detailRow_040867"
        >
          <p
            class="_font-body-sm-semibold_1jx6b_45 _detailLabel_040867"
          >
            Device ID
          </p>
          <p
            class="_font-body-sm-regular_1jx6b_40 _detailValue_040867"
          >
            <code>
              abcd1234
            </code>
          </p>
        </li>
        <li
          class="_detailRow_040867"
        >
          <p
            class="_font-body-sm-semibold_1jx6b_45 _detailLabel_040867"
          >
            Signed in
          </p>
          <p
            class="_font-body-sm-regular_1jx6b_40 _detailValue_040867"
          >
            <time
              datetime="2023-06-29T03:35:17Z"
            >
              Thu, 29 Jun 2023, 03:35
            </time>
          </p>
        </li>
      </ul>
    </div>
    <button
      aria-controls="radix-:r3:"
      aria-expanded="false"
      aria-haspopup="dialog"
      class="_button_1ukgx_17"
      data-kind="destructive"
      data-size="sm"
      data-state="closed"
      role="button"
      tabindex="0"
      type="button"
    >
      End session
    </button>
  </div>
</div>
`;
